From e1ba68cefd822b6670f7a95676e618eddd7882ec Mon Sep 17 00:00:00 2001
From: Mingcong Bai <jeffbai@aosc.io>
Date: Tue, 3 Dec 2024 16:52:14 +0800
Subject: [PATCH 190/301] AOSCOS: MIPS: loongson64: fix constant timer build on
 kernel versions >= v5.7-rc2

With commit 07d8350ede4c ("genirq: Remove setup_irq() and remove_irq()"),
setup_irq() was removed in favour of request_irq(). This was done
following commit cc2550b421aa ("clocksource: Replace setup_irq() by
request_irq()") (v5.7-rc1).

Follow the latter commit as a model to revise loongson64's constant_timer
and use request_irq() instead. Also add error handling and output for
cases of failure.

Signed-off-by: Mingcong Bai <jeffbai@aosc.io>
Signed-off-by: Kexy Biscuit <kexybiscuit@aosc.io>
---
 arch/mips/loongson64/constant_timer.c | 16 +++++++++-------
 1 file changed, 9 insertions(+), 7 deletions(-)

diff --git a/arch/mips/loongson64/constant_timer.c b/arch/mips/loongson64/constant_timer.c
index b8bdbb2c47b5..975fb93c4efa 100644
--- a/arch/mips/loongson64/constant_timer.c
+++ b/arch/mips/loongson64/constant_timer.c
@@ -42,12 +42,6 @@ static irqreturn_t constant_timer_interrupt(int irq, void *data)
 	return IRQ_NONE;
 }
 
-static struct irqaction constant_timer_irqaction = {
-	.handler = constant_timer_interrupt,
-	.flags = IRQF_PERCPU | IRQF_TIMER | IRQF_SHARED,
-	.name = "timer",
-};
-
 static int constant_set_state_oneshot(struct clock_event_device *evt)
 {
 	unsigned int raw_cpuid;
@@ -155,6 +149,7 @@ int constant_clockevent_init(void)
 	unsigned long max_delta = (1UL << 48) - 1;
 	unsigned int cpu = smp_processor_id();
 	struct clock_event_device *cd;
+	int err;
 
 	config = read_c0_config6();
 	config |= LOONGSON_CONF6_EXTIMER;
@@ -192,7 +187,14 @@ int constant_clockevent_init(void)
 		return 0;
 
 	constant_timer_irq_installed = 1;
-	setup_irq(irq, &constant_timer_irqaction);
+
+	err = request_irq(irq, constant_timer_interrupt,
+		    IRQF_PERCPU | IRQF_TIMER | IRQF_SHARED,
+		    "timer", cd);
+	if (err) {
+		pr_err("loongson64: setup irq for constant_clock_event failed: %d\n", err);
+		return err;
+	}
 
 	return 0;
 }
-- 
2.48.1

